lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py

```diff
@@ -195,257 +269,250 @@ def find_subnets(vpc_conn, vpc_id, identified_subnets):
     for table in route_tables:
-        this_tags = get_resource_tags(vpc_conn, table.id)
+        this_tags = describe_tags_with_backoff(connection, table['RouteTableId'])
         if tags_match(tags, this_tags):
             route_table = table
             count += 1
 
     if count > 1:
-        raise RuntimeError("Tags provided do not identify a unique route table")
+        module.fail_json(msg="Tags provided do not identify a unique route table")
     else:
         return route_table
 
 
 def route_spec_matches_route(route_spec, route):
-    key_attr_map = {
-        'destination_cidr_block': 'destination_cidr_block',
-        'gateway_id': 'gateway_id',
-        'instance_id': 'instance_id',
-        'interface_id': 'interface_id',
-        'vpc_peering_connection_id': 'vpc_peering_connection_id',
-    }
-
-    # This is a workaround to catch managed NAT gateways as they do not show
-    # up in any of the returned values when describing route tables.
-    # The caveat of doing it this way is that if there was an existing
-    # route for another nat gateway in this route table there is not a way to
-    # change to another nat gateway id. Long term solution would be to utilise
-    # boto3 which is a very big task for this module or to update boto.
-    if route_spec.get('gateway_id') and 'nat-' in route_spec['gateway_id']:
-        if route.destination_cidr_block == route_spec['destination_cidr_block']:
-            if all((not route.gateway_id, not route.instance_id, not route.interface_id, not route.vpc_peering_connection_id)):
-                return True
-
-    for k in key_attr_map:
-        if k in route_spec:
-            if route_spec[k] != getattr(route, k):
-                return False
-    return True
+    if route_spec.get('GatewayId') and 'nat-' in route_spec['GatewayId']:
+        route_spec['NatGatewayId'] = route_spec.pop('GatewayId')
+
+    return set(route_spec.items()).issubset(route.items())
 
 
 def route_spec_matches_route_cidr(route_spec, route):
-    cidr_attr = 'destination_cidr_block'
-    return route_spec[cidr_attr] == getattr(route, cidr_attr)
+    return route_spec['DestinationCidrBlock'] == route['DestinationCidrBlock']
 
 
 def rename_key(d, old_key, new_key):
-    d[new_key] = d[old_key]
-    del d[old_key]
+    d[new_key] = d.pop(old_key)
 
 
 def index_of_matching_route(route_spec, routes_to_match):
     for i, route in enumerate(routes_to_match):
         if route_spec_matches_route(route_spec, route):
-            return i
+            return "exact", i
         elif route_spec_matches_route_cidr(route_spec, route):
-            return "replace"
+            return "replace", i
 
 
-def ensure_routes(vpc_conn, route_table, route_specs, propagating_vgw_ids,
+def ensure_routes(connection, module, route_table, route_specs, propagating_vgw_ids,
                   check_mode, purge_routes):
-    routes_to_match = list(route_table.routes)
+    routes_to_match = [route for route in route_table['Routes']]
```

Comment:
Is `route_table['Routes']` a list already?

Reply:
Probably. I wonder if I was making a copy to avoid mutating the original route_table - I guess `routes_to_match = route_table['Routes'][:]` works as well in that case.